cmd/local-gadget/main.go

```diff
@@ -105,7 +105,11 @@ func newRootCmd() *cobra.Command {
 					fmt.Println(err.Error())
 					return
 				}
-				err = localGadgetManager.Operation(name, "start")
+				if gadget == "socket-collector" || gadget == "process-collector" {
+					err = localGadgetManager.Operation(name, "collect")
+				} else {
+					err = localGadgetManager.Operation(name, "start")
+				}
```

Comment:
I am not a fan of this part so I have two proposals:

1. Add a new method to the [TraceFactory interface](https://github.com/kinvolk/inspektor-gadget/blob/4a06ad3eb2ec53a321319c24dcc6e5ed7829ec8e/pkg/gadgets/interface.go#L28-L46) that returns the main operation. It would be mandatory for all the gadgets.
2. Add a new boolean flag to the [TraceOperation data structure](https://github.com/kinvolk/inspektor-gadget/blob/4a06ad3eb2ec53a321319c24dcc6e5ed7829ec8e/pkg/gadgets/interface.go#L60-L73) to identify which is the main operation. 
3. [Discarded] I thought also that we could use the [Order field](https://github.com/kinvolk/inspektor-gadget/blob/4a06ad3eb2ec53a321319c24dcc6e5ed7829ec8e/pkg/gadgets/interface.go#L70-L72) to enumerate the operations and the one with the value `1` would be the main operation but I don't like it.

This implementation could be also used by the trace-controller to implement the [Trace.Specs.RunMode=Auto](https://github.com/kinvolk/inspektor-gadget/blob/4a06ad3eb2ec53a321319c24dcc6e5ed7829ec8e/pkg/api/v1alpha1/trace_types.go#L50-L53).

I vote for option (1). What do you all think?

Reply:
I don't understand what you mean by "main" operation, the first operation that has to be executed? 

> Add a new method to the TraceFactory interface that returns the main operation. It would be mandatory for all the gadgets.

I think it's an overkill and keep adding more methods to that interface. 

> Add a new boolean flag to the TraceOperation data structure to identify which is the main operation.

Same as above. 

> I thought also that we could use the Order field to enumerate the operations and the one with the value 1 would be the main operation but I don't like it.

Why you don't like it? AFAIU the purpose of this field was to provide order for the operations, so I think using it to understand what's the first one is a good approach IMO. 

Also if we think specifically about process-collector and socket-collector they only support a single operation, so the caller could call the single element returned by `Operations()`

Reply:
> I don't understand what you mean by "main" operation, the first operation that has to be executed?
>
Yes, exactly. In general, I am trying to avoid hardcoding a `if gadget == "gadget-name"`.

> Why you don't like it?
> 
It is because every single operation for each gadget will have to specify the order (AFAIK, we are not currently doing it) and we will have to validate that it is 1, 2, ..., X , not duplicated, etc. Unless there is an automatic validation method like [this](https://github.com/kinvolk/inspektor-gadget/blob/4a06ad3eb2ec53a321319c24dcc6e5ed7829ec8e/pkg/api/v1alpha1/trace_types.go#L60) that I don't know 😅 .

> Also if we think specifically about process-collector and socket-collector they only support a single operation, so the caller could call the single element returned by Operations()
>
Could you please give more details about this? You mean that if no operation is specified then I could check if the number of operations is 1 and it that case I run that single operation? If so, I think I still have to add the `if gadget == "*-collector"` to send no operation for the socket/process-collector and "start" otherwise. Or what did I miss?

cc @alban

Reply:
> we will have to validate that it is 1, 2, ..., X , not duplicated, etc. Unless there is an automatic validation method like this that I don't know sweat_smile .

I don't think we need to worry about it. It's up to the developer to be sure those indexes are correct. 

> Could you please give more details about this? You mean that if no operation is specified then I could check if the number of operations is 1 and in that case I can run that single operation? If so, I think I still have to add the if gadget == "*-collector" to send no operation for the socket/process-collector and "start" otherwise. Or what did I miss?

I meant to have that logic on the local-gadget. Instead of having `localGadgetManager.Operation(name, "start")` there could be something like:

```go
operations = localGadgetManager.ListOperations(traceName)
if len(operations) == 1 {
    localGadgetManager.Operation(name, get_element(operations))
} else {
    localGadgetManager.Operation(name, "start")
}
```

This solution is still not ideal but (1) avoids adding more methods to the `TraceFactory` interface (2) hardcoding the name of the gadgets. 

I think the future solution is to implement `RunMode=auto` and use it.

Reply:
> I think the future solution is to implement RunMode=auto and use it.
>

I agree thus I applied your proposal. I don't want to work on this and maybe completely change it when we implement `RunMode=auto`.